src/EditorFeatures/CSharpTest/MakeFieldReadonly/MakeFieldReadonlyTests.cs

```diff
@@ -0,0 +1,748 @@
+@"class MyClass
+{
+    private static int [|_goo|];
+    static MyClass()
+    {
+        _goo = 0;
+    }
+}",
+@"class MyClass
+{
+    private static readonly int _goo;
+    static MyClass()
+    {
+        _goo = 0;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMakeFieldReadonly)]
+        public async Task StaticFieldAssignedInNonStaticCtor()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class MyClass
+{
+    private static int [|_goo|];
+    MyClass()
+    {
+        _goo = 0;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMakeFieldReadonly)]
+        public async Task FieldTypeIsMutableStruct()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"struct MyStruct
+{
+    private int _goo;
+}
+class MyClass
+{
+    private MyStruct [|_goo|];
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMakeFieldReadonly)]
+        public async Task FieldTypeIsCustomImmutableStruct()
+        {
+            await TestInRegularAndScriptAsync(
+@"struct MyStruct
+{
+    private readonly int _goo;
+    private const int _bar = 0;
+    private static int _fizz;
+}
+class MyClass
+{
+    private MyStruct [|_goo|];
+}",
+@"struct MyStruct
+{
+    private readonly int _goo;
+    private const int _bar = 0;
+    private static int _fizz;
+}
+class MyClass
+{
+    private readonly MyStruct _goo;
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMakeFieldReadonly)]
+        public async Task FixAll()
+        {
+            await TestInRegularAndScriptAsync(
+@"class MyClass
+{
+    private int {|FixAllInDocument:_goo|} = 0, _bar = 0;
+    private int _fizz = 0;
+}",
+@"class MyClass
+{
+    private readonly int _goo = 0, _bar = 0;
+    private readonly int _fizz = 0;
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMakeFieldReadonly)]
+        public async Task FixAll2()
+        {
+            await TestInRegularAndScriptAsync(
+@"  partial struct MyClass
+    {
+        private static Func<int, bool> {|FixAllInDocument:_test1|} = x => x > 0;
+        private static Func<int, bool> _test2 = x => x < 0;
+
+        private static Func<int, bool> _test3 = x =>
+        {
+            return x == 0;
+        };
+
+        private static Func<int, bool> _test4 = x =>
+        {
+            return x != 0;
+        };
+    }
+
+    partial struct MyClass { }",
+@"  partial struct MyClass
+    {
+        private static readonly Func<int, bool> _test1 = x => x > 0;
+        private static readonly Func<int, bool> _test2 = x => x < 0;
+
+        private static readonly Func<int, bool> _test3 = x =>
+        {
+            return x == 0;
+        };
+
+        private static readonly Func<int, bool> _test4 = x =>
+        {
+            return x != 0;
+        };
+    }
+
+    partial struct MyClass { }");
+        }
+    }
```

Comment:
i'd like a test where the field is written to from inside a nested type inside the class.  i.e. something like this:

```
class C
{
    private int i;

    class D : C
    {
        public D()
        {
            i++;
        }
    }
}
```

This should not be allowed to change even though 'i' is chnaged in a constructor.

Reply:
➡️ This was implemented with tests in f19cceb66030ae14c9f11850b08d3ac9694059bd